content/posts/codegen.md

```diff
@@ -0,0 +1,143 @@
+we've chosen not to. We actually take a third approach: templating.
+
+## Templated Approach
+
+The approach we take is essentially Mad Libs.
+
+<img src="/codegen.png" class="img-fluid hero-image" alt="Codegen example">
+
+You have a text file that looks something like code. At the same time, you have
+a pile of parameters. Your code generator is responsible for taking parameters,
+injecting them into your code template, and then writing the resulting file.
+It's Mad Libs!
+
+We primarly use templates for ease of iteration. We can make most changes to the
+templates easily and be reasonably certain that the resulting code will appear
+as we intend. Refactoring the generated code can be done at the template level
+and doesn't require large generator rewrites. We can also programatically
+generate anything we want. We can use the same sets of primatives and
+infrastructure to create Markdown documentation, Go code, Python code,
+protobufs, YAML files and anything in between. 
+
+The downside to this approach is that the templates can grow unwiedy. As we
+discover more places where customizations need to occur, those customizations
+become embedded into our templates. IDEs have poor support for understanding
+template syntax, leaving it up to the team to carefully parse our templates and
+understand what's happening.
+
+## Abstractions: When to Generate?
+
+Typically, when writing code, we think about abstractions. We start out with a
+list of instructions. We can create functions, which abstract away those list
+of instructions. Then, we can create classes, which abstract away both the
+functions and the underlying data used by the functions. Code generation acts
+as a layer on top of all of these.
+
+Because of this, we're constantly asking ourselves if we're using the right
+abstraction. Human-written code is easy to test. Generated code is easy to
+test, but the generator itself is hard to test. This means that we're always on
+the hunt for more edge cases in generated code, so we can improve the generator
+better.
+
+I suspect that this is why using "compiler magic" is so much more tenable for
+teams. It's easier to write unit tests when you have handfuls of functions that
+all alter an AST. That's how developers learn to write code! Write a function
+and then write a text for that function. A template approach means that most of
+the true "logic" of your program is embedded inside a templating library that
+you have no insight into. You can test the end result, but you can't easily
+inject tests into anything in the middle.
+
+Our team does rewrites where we take a large amount of template, write a
+function that replicates that template at runtime, and then replace the
+template with a single line of template code. We can more easily test our
+handwritten function. At that point, we just have to make sure that the
+template places the correct values into the function.
```

Comment:
Love this section. I think it doesn't feel like a conclusion though, it feels like it abruptly ends here. Might suggest writing some sort of conclusion or something!